### PR TITLE
fix(gatsby): Move "react-dom-server" out of framework chunk

### PR DESCRIPTION
**Pull request by LekoArts**
## Description

When one uses `gatsby-plugin-image` the `react-dom-server` is included into the framework chunk.

Before:

image

After:

image

## Related Issues

[ch60411]

Fixes https://github.com/gatsbyjs/gatsby/issues/37281

### CODE REVIEW

File: packages/gatsby/src/utils/webpack.config.js [hunk 1]
```diff
@@ -26,10 +26,16 @@ import { shouldGenerateEngines } from "./engines-helpers"
 import { ROUTES_DIRECTORY } from "../constants"
 import { BabelConfigItemsCacheInvalidatorPlugin } from "./babel-loader"
 import { PartialHydrationPlugin } from "./webpack/plugins/partial-hydration"
-import { satisfiesSemvers } from "./flags"
 
 const FRAMEWORK_BUNDLES = [`react`, `react-dom`, `scheduler`, `prop-types`]
 
+// This regex ignores nested copies of framework libraries so they're bundled with their issuer
+const FRAMEWORK_BUNDLES_REGEX = new RegExp(
```
Comment (LekoArts) [inline packages/gatsby/src/utils/webpack.config.js:33]: Moving regex creation to top. Better performance than creating the regex over and over again

File: packages/gatsby/src/utils/webpack.config.js [hunk 2]
```diff
@@ -582,12 +588,7 @@ module.exports = async (
           framework: {
             chunks: `all`,
             name: `framework`,
-            // This regex ignores nested copies of framework libraries so they're bundled with their issuer.
-            test: new RegExp(
-              `(?<!node_modules.*)[\\\\/]node_modules[\\\\/](${FRAMEWORK_BUNDLES.join(
-                `|`
-              )})[\\\\/]`
-            ),
+            test: FRAMEWORK_BUNDLES_REGEX,
```
Comment (LekoArts) [inline packages/gatsby/src/utils/webpack.config.js:591]: In develop stage we don't need to change anything, `react-dom/server` is only relevant for `build-javascript` stage

File: packages/gatsby/src/utils/webpack.config.js [hunk 3]
```diff
@@ -634,12 +635,21 @@ module.exports = async (
...
+            // - "node_modules/react-dom/cjs/react-dom-server-legacy.browser.production.min.js"
+            // Use a "/" before "react-dom-server" so that we don't match packages that contain "react-dom-server" in their name
+            if (
+              module.rawRequest === `react-dom/server` ||
+              module.rawRequest.includes(`/react-dom-server`)
+            ) {
+              return false
+            }
+
+            return FRAMEWORK_BUNDLES_REGEX.test(module.nameForCondition())
```
Comment (LekoArts) [inline packages/gatsby/src/utils/webpack.config.js:651]: We seem to use `module.identifier()` in other places but most often on the internet I've seen people use `module.nameForCondition()`

Comment (wardpeet) [reply packages/gatsby/src/utils/webpack.config.js:651]: it doesn't really matter for this use case. nameForCondition returns the resource without any query parts.
https://github.com/webpack/webpack/blob/293e677b355da0d5bc4ddfc97d2afec114912711/lib/NormalModule.js#L375-L380

**Review by wardpeet — APPROVED** [review]

File: packages/gatsby/src/utils/webpack.config.js [hunk 4]
```diff
@@ -634,12 +635,21 @@ module.exports = async (
...
-          ),
+          test: module => {
+            // Packages like gatsby-plugin-image might import from "react-dom/server". We don't want to include react-dom-server in the framework bundle.
+            // A rawRequest might look like these:
+            // - "react-dom/server"
+            // - "node_modules/react-dom/cjs/react-dom-server-legacy.browser.production.min.js"
+            // Use a "/" before "react-dom-server" so that we don't match packages that contain "react-dom-server" in their name
+            if (
+              module.rawRequest === `react-dom/server` ||
+              module.rawRequest.includes(`/react-dom-server`)
```
[the remark this answers is not in the document]
Comment (LekoArts) [reply packages/gatsby/src/utils/webpack.config.js]: @wardpeet do you have an opinion here?

[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby/src/utils/webpack.config.js]: rawRequest as that's the actual user's request. If you use request, it's appended with loaders etc which  will be harder to match

File: packages/gatsby/src/utils/webpack.config.js (hunk 3 above)
Comment (amalitsky) [reply packages/gatsby/src/utils/webpack.config.js:651]: this change brakes _gatsby-plugin-preact_ which [assumes that `test` function is a regex](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-preact/src/gatsby-node.js#L56)

**Pull request merged**